src/libAtomVM/externalterm.c

```diff
@@ -439,7 +441,39 @@ static term parse_external_terms(const uint8_t *external_term_buf, size_t *eterm
         case ATOM_EXT: {
             uint16_t atom_len = READ_16_UNALIGNED(external_term_buf + 1);
 
-            int global_atom_id = globalcontext_insert_atom_maybe_copy(glb, (AtomString) (external_term_buf + 2), copy);
+            if (UNLIKELY(atom_len > 255)) {
+                return term_invalid_term();
+            }
+
+            const uint8_t *atom_chars = (const uint8_t *) (external_term_buf + 3);
+            int global_atom_id;
+            if (LIKELY(unicode_buf_is_ascii(atom_chars, atom_len))) {
+                // there is a trick here: we are reusing LSB of len field as atom length
+                global_atom_id = globalcontext_insert_atom_maybe_copy(
+                    glb, (AtomString) (external_term_buf + 2), copy);
+            } else {
+                // need to re-encode latin1 to UTF-8
+                size_t required_buf_size = unicode_latin1_buf_size_as_utf8(atom_chars, atom_len);
+                if (UNLIKELY(required_buf_size > 255)) {
+                    return term_invalid_term();
+                }
+                uint8_t *atom_buf = malloc(1 + required_buf_size);
+                atom_buf[0] = required_buf_size;
+                uint8_t *curr_codepoint = atom_buf;
+                for (int i = 0; i < atom_len; i++) {
+                    size_t codepoint_size;
+                    // latin1 encoding is always successful
+                    bitstring_utf8_encode(atom_chars[i], curr_codepoint, &codepoint_size);
+                    curr_codepoint += codepoint_size;
+                }
+                global_atom_id
+                    = globalcontext_insert_atom_maybe_copy(glb, (AtomString) atom_buf, true);
+                free(atom_buf);
+            }
+
```

Comment:
When a new atom is inserted to the atom table an allocation may happen, in case of failure, a negative value is returned.

Reply:
I get it.  I guess in the future we would want to change the signature of this function so we can return an atom that indicates that this is an allocation failure, rather than something the user did, but that is for another day.